server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java

```diff
@@ -62,198 +67,334 @@ public class Upgrader11to12Test {
 
-    replay(bw);
+    LOG.info("c:{}", capturedUpdate2.getValue().prettyPrint());
+    var u2 = capturedUpdate2.getValue();
+    // 1 add, 1 delete
+    assertEquals(2, u2.getUpdates().size());
 
-    String fileName = "hdfs://localhost:8020/accumulo/tables/12/default_tablet/A000000v.rf";
-    Key k = Key.builder().row(new Text("12;")).family(DataFileColumnFamily.NAME)
-        .qualifier(new Text(fileName)).build();
-    Value v = new Value("1234,5678");
+    verify(batchWriter, scanner);
 
-    assertThrows(IllegalStateException.class, () -> upgrader.upgradeDataFileCF(k, v, bw, "aTable"));
-
-    verify(bw);
   }
 
   @Test
-  void upgradeDataFileCFInvalidPathTest() {
-    Upgrader11to12 upgrader = new Upgrader11to12();
-
-    BatchWriter bw = createMock(BatchWriter.class);
-
-    replay(bw);
+  public void skipConvertedFileTest() throws Exception {
+    BatchWriter batchWriter = mock(BatchWriter.class);
+    Capture<Mutation> capturedUpdate1 = newCapture();
+    batchWriter.addMutation(capture(capturedUpdate1));
+    expectLastCall().once();
+    // create sample data "served" by the mocked scanner
+    TreeMap<Key,Value> scanData = new TreeMap<>();
+    Text row1 = new Text("123");
+
+    // reference already in expected form with fence info.
+    String fileName1 =
+        "{\"path\":\"hdfs://localhost:8020/accumulo/tables/12/default_tablet/A000000v.rf\",\"startRow\":\"\",\"endRow\":\"\"}";
+    Key key1 =
+        Key.builder(false).row(row1).family(DataFileColumnFamily.NAME).qualifier(fileName1).build();
+    Value value1 = new Value("123,456");
+    scanData.put(key1, value1);
+
+    String fileName2 = "hdfs://localhost:8020/accumulo/tables/12/default_tablet/B000000v.rf";
+    Key key2 =
+        Key.builder(false).row(row1).family(DataFileColumnFamily.NAME).qualifier(fileName2).build();
+    Value value2 = new Value("321,654");
+    scanData.put(key2, value2);
+
+    Scanner scanner = mock(Scanner.class);
+    scanner.fetchColumnFamily(DataFileColumnFamily.NAME);
+    expectLastCall();
+    scanner.fetchColumnFamily(ChoppedColumnFamily.NAME);
+    expectLastCall();
+    scanner.fetchColumnFamily(ExternalCompactionColumnFamily.NAME);
+    expectLastCall();
 
-    String invalidPath = "badPath";
+    expect(scanner.iterator()).andReturn(scanData.entrySet().iterator()).once();
+    replay(batchWriter, scanner);
 
-    Key k = Key.builder().row(new Text("12;")).family(DataFileColumnFamily.NAME)
-        .qualifier(new Text(invalidPath)).build();
-    Value v = new Value("1234,5678");
+    Upgrader11to12 upgrader = new Upgrader11to12();
+    upgrader.processReferences(batchWriter, scanner, "accumulo.metadata");
 
-    assertThrows(IllegalArgumentException.class,
-        () -> upgrader.upgradeDataFileCF(k, v, bw, "aTable"));
+    LOG.info("c:{}", capturedUpdate1.getValue().prettyPrint());
+    var u1 = capturedUpdate1.getValue();
+    // 1 add, 1 delete
+    assertEquals(2, u1.getUpdates().size());
```

Comment:
could also stream and count deletes here

Reply:
Added with e09da28032